returnn/tf/layers/basic.py

```diff
@@ -909,11 +955,31 @@ def get_out_data_from_opts(cls, name, sources=(), start=None, size=None, **kwarg
     :rtype: Data
     """
     from ..util.data import DimensionTag
-    input_data = get_concat_sources_data_template(sources).copy_as_batch_spatial_major()
-    if start:
-      input_data.beam = SearchBeam.get_combined_beam(input_data.beam, start.output.beam)
-    new_dim_tag = DimensionTag(kind=DimensionTag.Types.Spatial, description="%s:slice_nd" % name, dimension=size)
-    return input_data.copy_template_replace_dim_tag(axis=1, new_dim_tag=new_dim_tag, name="%s_output" % name)
+    start_data = start.output.copy_as_batch_major()
+    input_data = sources[0].output.copy_as_batch_major()
+    input_t = input_data.placeholder
+    indices_data = start_data.copy_template(name="%s_gather_indices" % name)
+    if isinstance(size, int):
+      tag = DimensionTag(
+        kind=DimensionTag.Types.Spatial,
+        description="time_sliced",
+        dimension=size)
+    else:
+      # get tensor of shape (B,) from input to use as dynamic size (start_data has no placeholder here)
+      dyn_size = tf.repeat(tf.reduce_max(input_t), tf.shape(input_t)[0])
+      dyn_size = tf.cast(dyn_size, tf.int32)
```

Comment:
This is in the `get_out_data_from_opts` function, where no TF operations should be done.
Just leave the `dyn_size` away here.

Reply:
If I exclude `dyn_size` completely, I get an error in the case of `size=None`. I think either `dimension` or `dyn_size` needs to be set, right?

Reply:
`dimension=None` is the default. But you might want to add it explicitly to make it more clear.
I don't think you get an error here. Of course you need to set it later (where I commented).

Reply:
So I just tried setting `optimize_move_layers_out = False` in my config and this leads to an error since we removed the `dyn_size` from `get_out_data_from_opts`.

The network is basically the same as for the issue I posted a few days ago:

```
network = {
  "output": {
    "class": "rec", "from": "data:data",  # time-sync
    "include_eos": True, "back_prop": (task == "train") and train, "unit": {
      "segment_starts": {  # (B,)
        "class": "switch", "condition": "prev:output_is_not_blank", "true_from": ":i",
        "false_from": "prev:segment_starts", "initial_output": 0
      },

      "segments": {"class": "slice_nd", "from": "base:data:data", "start": "segment_starts", "size": None,
      "is_output_layer": True},

      "output_prob": {
        "class": "softmax", "from": "data:source", "target": "classes", "loss": "ce"
      },
      'output': {
        'class': 'choice', 'target': "classes", 'beam_size': 3, 'from': "output_prob", "input_type": "prob",
        "initial_output": 0,
      }, "output_is_not_blank": {
        "class": "compare", "from": "output", "value": 0, "kind": "not_equal", "initial_output": True
      },
    },
  }
}
```

I get the following error:

```
/home/robin/miniconda3/envs/i6_env/bin/python3 /mnt/projects/i6/returnn/rnn.py /mnt/projects/i6/experiments/transducer/config/configs/segments-tests/error_reconstruct2.py
2021-09-14 13:52:05.722700: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2021-09-14 13:52:05.722716: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
RETURNN starting up, version 1.0.0+unknown, date/time 2021-09-14-13-52-06 (UTC+0200), pid 26443, cwd /mnt/projects/i6/returnn, Python /home/robin/miniconda3/envs/i6_env/bin/python3
RETURNN command line options: ['/mnt/projects/i6/experiments/transducer/config/configs/segments-tests/error_reconstruct2.py']
Hostname: robin-MS-7B86
TensorFlow: 2.5.0 (v2.5.0-rc3-213-ga4dfb8d1a71) (<site-package> in /home/robin/miniconda3/envs/i6_env/lib/python3.9/site-packages/tensorflow)
Setup TF inter and intra global thread pools, num_threads None, session opts {'log_device_placement': False, 'device_count': {'GPU': 0}}.
2021-09-14 13:52:06.659799: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-09-14 13:52:06.660864: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcuda.so.1
CUDA_VISIBLE_DEVICES is not set.
Collecting TensorFlow device list...
2021-09-14 13:52:06.702726: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-09-14 13:52:06.702746: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]      
2021-09-14 13:52:06.703797: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-09-14 13:52:06.704190: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1733] Found device 0 with properties: 
pciBusID: 0000:1f:00.0 name: NVIDIA GeForce GTX 1660 SUPER computeCapability: 7.5
coreClock: 1.8GHz coreCount: 22 deviceMemorySize: 5.80GiB deviceMemoryBandwidth: 312.97GiB/s
2021-09-14 13:52:06.704259: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2021-09-14 13:52:06.704296: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcublas.so.11'; dlerror: libcublas.so.11: cannot open shared object file: No such file or directory
2021-09-14 13:52:06.704331: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcublasLt.so.11'; dlerror: libcublasLt.so.11: cannot open shared object file: No such file or directory
2021-09-14 13:52:06.704366: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcufft.so.10'; dlerror: libcufft.so.10: cannot open shared object file: No such file or directory
2021-09-14 13:52:06.704400: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcurand.so.10'; dlerror: libcurand.so.10: cannot open shared object file: No such file or directory
2021-09-14 13:52:06.704433: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcusolver.so.11'; dlerror: libcusolver.so.11: cannot open shared object file: No such file or directory
2021-09-14 13:52:06.704465: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcusparse.so.11'; dlerror: libcusparse.so.11: cannot open shared object file: No such file or directory
2021-09-14 13:52:06.704501: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudnn.so.8'; dlerror: libcudnn.so.8: cannot open shared object file: No such file or directory
2021-09-14 13:52:06.704508: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1766] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2021-09-14 13:52:06.790980: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-09-14 13:52:06.791002: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]      0 
2021-09-14 13:52:06.791008: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1277] 0:   N 
Local devices available to TensorFlow:
  1/1: name: "/device:CPU:0"
       device_type: "CPU"
       memory_limit: 268435456
       locality {
       }
       incarnation: 2835650696115532840
Train data:
  input: 10 x 1
  output: {'classes': (3, 1), 'data': (10, 2)}
  DummyDataset, sequences: 10, frames: unknown
Device not set explicitly, and no GPU found.
Learning-rate-control: no file specified, not saving history (no proper restart possible)
Setup TF session with options {'log_device_placement': False, 'device_count': {'GPU': 0}} ...
2021-09-14 13:52:06.861157: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-09-14 13:52:06.861192: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]      
Warning: assuming dim tags are same with different size placeholders: DimensionTag{'time'[B?]} vs DimensionTag{'time'[B?]}
layer root/'data:data' output: Data{'data', [B,T|'time'[B],F|F'feature:data'(10)]}
WARNING:tensorflow:From /mnt/projects/i6/returnn/returnn/tf/network.py:465: calling Zeros.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
layer root/'output' output: Data{'output_output', [T|'time'[B],B], dtype='int32', sparse=True, dim=3}
WARNING:tensorflow:From /mnt/projects/i6/returnn/returnn/tf/util/basic.py:1343: calling VarianceScaling.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Exception occurred during in-loop construction of layer 'classes'.
We had previous exceptions at template construction, which got resolved, but maybe sth is wrong.
Template network (check out types / shapes):
{':i': <_TemplateLayer(RecStepInfoLayer)(:template::i) output/':i' out_type=Data{[], dtype='int32', ctx=loop('time'[B])} (construction stack 'segment_starts')>,
 'data:classes': <_TemplateLayer(SourceLayer)(:template:source) output/'data:classes' out_type=Data{[B], dtype='int32', sparse=True, dim=3, available_for_inference=False, ctx=loop('time'[B])} (construction stack 'output')>,
 'data:source': <_TemplateLayer(SourceLayer)(:template:source) output/'data:source' out_type=Data{[B,F|F'feature:data'(10)], ctx=loop('time'[B])} (construction stack 'output_prob')>,
 'output': <_TemplateLayer(ChoiceLayer)(:template:choice) output/'output' out_type=Data{[B], dtype='int32', sparse=True, dim=3, ctx=loop('time'[B])} (construction stack None)>,
 'output_is_not_blank': <_TemplateLayer(CompareLayer)(:template:compare) output/'output_is_not_blank' out_type=Data{[B], dtype='bool', sparse=True, dim=2, ctx=loop('time'[B])} (construction stack 'segment_starts')>,
 'output_prob': <_TemplateLayer(SoftmaxLayer)(:template:softmax) output/'output_prob' out_type=Data{[B,F|F'output_prob:feature-dense'(3)], ctx=loop('time'[B])} (construction stack None)>,
 'segment_starts': <_TemplateLayer(SwitchLayer)(:template:switch) output/'segment_starts' out_type=Data{[B], dtype='int32', ctx=loop('time'[B])} (construction stack 'segments')>,
 'segments': <_TemplateLayer(SliceNdLayer)(:template:slice_nd) output/'segments' out_type=Data{[B,T|'sliced-time:segments'[?],F|F'feature:data'(10)], ctx=loop('time'[B])} (construction stack None)>}
Collected (unique) exceptions during template construction:
(Note that many of these can be ignored, or are expected.)
EXCEPTION while constructing layer 'segment_starts'
NetworkConstructionDependencyLoopException: <TFNetwork 'root/output(rec-subnet)' parent_net=<TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>: Error: There is a dependency loop on layer 'segment_starts'.
Construction stack (most recent first):
  segment_starts
  segments

layer root/output(rec-subnet)/'data:classes' output: Data{'classes', [B], dtype='int32', sparse=True, dim=3, available_for_inference=False, ctx=loop('time'[B])}
layer root/output(rec-subnet)/'output' output: Data{'output_output', [B], dtype='int32', sparse=True, dim=3, ctx=loop('time'[B])}
layer root/output(rec-subnet)/'data:source' output: Data{'data', [B,F|F'feature:data'(10)], ctx=loop('time'[B])}
Exception occurred during in-loop construction of layer 'classes'.
layer root/output(rec-subnet)/'output_prob' output: Data{'output_prob_output', [B,F|F'output_prob:feature-dense'(3)], ctx=loop('time'[B])}
layer root/output(rec-subnet)/'segment_starts' output: Data{'segment_starts_output', [B], dtype='int32', ctx=loop('time'[B])}
layer root/output(rec-subnet)/'segments' output: Data{'segments_gather_output', [B,T|'sliced-time:segments'[?],F|F'feature:data'(10)], ctx=loop('time'[B])}
WARNING:tensorflow:From /home/robin/miniconda3/envs/i6_env/lib/python3.9/site-packages/tensorflow/python/ops/array_ops.py:5043: calling gather (from tensorflow.python.ops.array_ops) with validate_indices is deprecated and will be removed in a future version.
Instructions for updating:
The `validate_indices` argument has no effect. Indices are always validated on CPU and never validated on GPU.
layer root/output(rec-subnet)/'output_is_not_blank' output: Data{'output_is_not_blank_output', [B], dtype='bool', sparse=True, dim=2, ctx=loop('time'[B])}
layer root/output(rec-subnet-output)/'segments' output: Data{'segments_gather_output', [T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]}
Exception creating layer root/output(rec-subnet-output)/'segments' of class InternalLayer with opts:
{'name': 'segments',
 'network': <TFNetwork 'root/output(rec-subnet-output)' parent_layer=<RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>,
 'output': Data{'segments_gather_output', [T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]},
 'sources': []}
Exception creating layer root/'output' of class RecLayer with opts:
{'_name': 'output',
 '_network': <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>,
 '_time_dim_tag': DimensionTag{'time'[B]},
 'back_prop': {'class': 'DummyDataset',
               'input_dim': 10,
               'num_seqs': 10,
               'output_dim': 3,
               'seq_len': 10},
 'include_eos': True,
 'n_out': <class 'returnn.util.basic.NotSpecified'>,
 'name': 'output',
 'network': <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>,
 'output': Data{'output_output', [T|'time'[B],B], dtype='int32', sparse=True, dim=3},
 'sources': [<SourceLayer 'data:data' out_type=Data{[B,T|'time'[B],F|F'feature:data'(10)]}>],
 'unit': <_SubnetworkRecCell 'root/output(rec-subnet)'>}
Unhandled exception <class 'AssertionError'> in thread <_MainThread(MainThread, started 140427022303616)>, proc 26443.

Thread current, main, <_MainThread(MainThread, started 140427022303616)>:
(Excluded thread.)

That were all threads.
WARNING:tensorflow:From /mnt/projects/i6/returnn/returnn/tf/util/basic.py:5445: _colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
EXCEPTION
Traceback (most recent call last):
  File "/mnt/projects/i6/returnn/rnn.py", line 11, in <module>
    line: main()
    locals:
      main = <local> <function main at 0x7fb75121a940>
  File "/mnt/projects/i6/returnn/returnn/__main__.py", line 655, in main
    line: execute_main_task()
    locals:
      execute_main_task = <global> <function execute_main_task at 0x7fb75121a820>
  File "/mnt/projects/i6/returnn/returnn/__main__.py", line 453, in execute_main_task
    line: engine.init_train_from_config(config, train_data, dev_data, eval_data)
    locals:
      engine = <global> <returnn.tf.engine.Engine object at 0x7fb751239fd0>
      engine.init_train_from_config = <global> <bound method Engine.init_train_from_config of <returnn.tf.engine.Engine object at 0x7fb751239fd0>>
      config = <global> <returnn.config.Config object at 0x7fb755914820>
      train_data = <global> <DummyDataset 'train' epoch=None>
      dev_data = <global> None
      eval_data = <global> None
  File "/mnt/projects/i6/returnn/returnn/tf/engine.py", line 1031, in Engine.init_train_from_config
    line: self.init_network_from_config(config)
    locals:
      self = <local> <returnn.tf.engine.Engine object at 0x7fb751239fd0>
      self.init_network_from_config = <local> <bound method Engine.init_network_from_config of <returnn.tf.engine.Engine object at 0x7fb751239fd0>>
      config = <local> <returnn.config.Config object at 0x7fb755914820>
  File "/mnt/projects/i6/returnn/returnn/tf/engine.py", line 1096, in Engine.init_network_from_config
    line: self._init_network(net_desc=net_dict, epoch=self.epoch)
    locals:
      self = <local> <returnn.tf.engine.Engine object at 0x7fb751239fd0>
      self._init_network = <local> <bound method Engine._init_network of <returnn.tf.engine.Engine object at 0x7fb751239fd0>>
      net_desc = <not found>
      net_dict = <local> {'enc_ctx_win': {'class': 'window', 'from': 'data:data', 'window_size': 5}, 'output': {'class': 'rec', 'from': 'data:data', 'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, 'unit': {'segment_starts': {'class': 'switch', ...
      epoch = <local> None
      self.epoch = <local> 1
  File "/mnt/projects/i6/returnn/returnn/tf/engine.py", line 1275, in Engine._init_network
    line: self.network, self.updater = self.create_network(
            config=self.config,
            extern_data=extern_data,
            rnd_seed=net_random_seed,
            train_flag=train_flag, eval_flag=self.use_eval_flag, search_flag=self.use_search_flag,
            initial_learning_rate=getattr(self, "initial_learning_rate", None),
            net_dict=net_desc)
    locals:
      self = <local> <returnn.tf.engine.Engine object at 0x7fb751239fd0>
      self.network = <local> None
      self.updater = <local> None
      self.create_network = <local> <bound method Engine.create_network of <class 'returnn.tf.engine.Engine'>>
      config = <not found>
      self.config = <local> <returnn.config.Config object at 0x7fb755914820>
      extern_data = <local> <ExternData data={'data': Data{'data', [B,T|'time'[B],F|F'feature:data'(10)]}, 'classes': Data{'classes', [B,T|'time'[B]], dtype='int32', sparse=True, dim=3, available_for_inference=False}}>
      rnd_seed = <not found>
      net_random_seed = <local> 1
      train_flag = <local> <tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>
      eval_flag = <not found>
      self.use_eval_flag = <local> True
      search_flag = <not found>
      self.use_search_flag = <local> False
      initial_learning_rate = <not found>
      getattr = <builtin> <built-in function getattr>
      net_dict = <not found>
      net_desc = <local> {'enc_ctx_win': {'class': 'window', 'from': 'data:data', 'window_size': 5}, 'output': {'class': 'rec', 'from': 'data:data', 'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, 'unit': {'segment_starts': {'class': 'switch', ...
  File "/mnt/projects/i6/returnn/returnn/tf/engine.py", line 1316, in Engine.create_network
    line: network.construct_from_dict(net_dict)
    locals:
      network = <local> <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      network.construct_from_dict = <local> <bound method TFNetwork.construct_from_dict of <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      net_dict = <local> {'enc_ctx_win': {'class': 'window', 'from': 'data:data', 'window_size': 5}, 'output': {'class': 'rec', 'from': 'data:data', 'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, 'unit': {'segment_starts': {'class': 'switch', ...
  File "/mnt/projects/i6/returnn/returnn/tf/network.py", line 605, in TFNetwork.construct_from_dict
    line: self.construct_layer(net_dict, name, get_layer=get_layer)
    locals:
      self = <local> <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      self.construct_layer = <local> <bound method TFNetwork.construct_layer of <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      net_dict = <local> {'enc_ctx_win': {'class': 'window', 'from': 'data:data', 'window_size': 5}, 'output': {'class': 'rec', 'from': 'data:data', 'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, 'unit': {'segment_starts': {'class': 'switch', ...
      name = <local> 'output', len = 6
      get_layer = <local> None
  File "/mnt/projects/i6/returnn/returnn/tf/network.py", line 931, in TFNetwork.construct_layer
    line: return add_layer(name=name_with_prefix, layer_class=layer_class, **layer_desc)
    locals:
      add_layer = <local> <bound method TFNetwork.add_layer of <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      name = <local> 'output', len = 6
      name_with_prefix = <local> 'output', len = 6
      layer_class = <local> <class 'returnn.tf.layers.rec.RecLayer'>
      layer_desc = <local> {'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, '_network': <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>, '_name': 'output', 'n_out': <class 'returnn.util.basic.NotSpecified'>, 'sourc..., len = 8
  File "/mnt/projects/i6/returnn/returnn/tf/network.py", line 1077, in TFNetwork.add_layer
    line: layer = self._create_layer(name=name, layer_class=layer_class, **layer_desc)
    locals:
      layer = <not found>
      self = <local> <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      self._create_layer = <local> <bound method TFNetwork._create_layer of <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      name = <local> 'output', len = 6
      layer_class = <local> <class 'returnn.tf.layers.rec.RecLayer'>
      layer_desc = <local> {'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, '_network': <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>, '_name': 'output', 'n_out': <class 'returnn.util.basic.NotSpecified'>, 'sourc..., len = 8
  File "/mnt/projects/i6/returnn/returnn/tf/network.py", line 998, in TFNetwork._create_layer
    line: layer = layer_class(**layer_desc)
    locals:
      layer = <not found>
      layer_class = <local> <class 'returnn.tf.layers.rec.RecLayer'>
      layer_desc = <local> {'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, '_network': <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>, '_name': 'output', 'n_out': <class 'returnn.util.basic.NotSpecified'>, 'sourc..., len = 11
  File "/mnt/projects/i6/returnn/returnn/tf/layers/rec.py", line 238, in RecLayer.__init__
    line: y = self._get_output_subnet_unit(self.cell)
    locals:
      y = <not found>
      self = <local> <RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}>
      self._get_output_subnet_unit = <local> <bound method RecLayer._get_output_subnet_unit of <RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}>>
      self.cell = <local> <_SubnetworkRecCell 'root/output(rec-subnet)'>
  File "/mnt/projects/i6/returnn/returnn/tf/layers/rec.py", line 925, in RecLayer._get_output_subnet_unit
    line: output = cell.get_output()
    locals:
      output = <not found>
      cell = <local> <_SubnetworkRecCell 'root/output(rec-subnet)'>
      cell.get_output = <local> <bound method _SubnetworkRecCell.get_output of <_SubnetworkRecCell 'root/output(rec-subnet)'>>
  File "/mnt/projects/i6/returnn/returnn/tf/layers/rec.py", line 2550, in _SubnetworkRecCell.get_output
    line: self._construct_output_layers_moved_out(
            loop_accumulated=self.final_acc_tas_dict, seq_len=seq_len,
            extra_output_layers=extra_output_layers, final_net_vars=final_net_vars)
    locals:
      self = <local> <_SubnetworkRecCell 'root/output(rec-subnet)'>
      self._construct_output_layers_moved_out = <local> <bound method _SubnetworkRecCell._construct_output_layers_moved_out of <_SubnetworkRecCell 'root/output(rec-subnet)'>>
      loop_accumulated = <not found>
      self.final_acc_tas_dict = <local> {'loss_output_prob': <tf.TensorArray 'output/rec/subnet_base/acc_ta_loss_output_prob'>, 'error_output_prob': <tf.TensorArray 'output/rec/subnet_base/acc_ta_error_output_prob'>, 'output_output': <tf.TensorArray 'output/rec/subnet_base/acc_ta_output_output'>, 'output_segments': <tf.TensorArray 'out...
      seq_len = <local> <tf.Tensor 'output/rec/subnet_base/check_seq_len_batch_size/check_input_dim/identity_with_dim_check:0' shape=(?,) dtype=int32>
      extra_output_layers = <local> {'segments', 'output'}, len = 2
      final_net_vars = <local> ([<tf.Tensor 'output/rec/while/Exit_1:0' shape=(?,) dtype=bool>, <tf.Tensor 'output/rec/while/Exit_2:0' shape=(?,) dtype=int32>], [])
  File "/mnt/projects/i6/returnn/returnn/tf/layers/rec.py", line 3311, in _SubnetworkRecCell._construct_output_layers_moved_out
    line: self.output_layers_net.layers[layer_name] = get_layer(layer_name)
    locals:
      self = <local> <_SubnetworkRecCell 'root/output(rec-subnet)'>
      self.output_layers_net = <local> <TFNetwork 'root/output(rec-subnet-output)' parent_layer=<RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      self.output_layers_net.layers = <local> {}
      layer_name = <local> 'segments', len = 8
      get_layer = <local> <function _SubnetworkRecCell._construct_output_layers_moved_out.<locals>.get_layer at 0x7fb7b3831a60>
  File "/mnt/projects/i6/returnn/returnn/tf/layers/rec.py", line 3303, in _SubnetworkRecCell._construct_output_layers_moved_out.<locals>.get_layer
    line: return get_loop_acc_layer(name)
    locals:
      get_loop_acc_layer = <local> <function _SubnetworkRecCell._construct_output_layers_moved_out.<locals>.get_loop_acc_layer at 0x7fb7b37b2940>
      name = <local> 'segments', len = 8
  File "/mnt/projects/i6/returnn/returnn/tf/layers/rec.py", line 3244, in _SubnetworkRecCell._construct_output_layers_moved_out.<locals>.get_loop_acc_layer
    line: layer_ = self.output_layers_net.add_layer(
            name=name, output=output, layer_class=InternalLayer, sources=[])
    locals:
      layer_ = <not found>
      self = <local> <_SubnetworkRecCell 'root/output(rec-subnet)'>
      self.output_layers_net = <local> <TFNetwork 'root/output(rec-subnet-output)' parent_layer=<RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      self.output_layers_net.add_layer = <local> <bound method TFNetwork.add_layer of <TFNetwork 'root/output(rec-subnet-output)' parent_layer=<RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      name = <local> 'segments', len = 8
      output = <local> Data{'segments_gather_output', [T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]}
      layer_class = <not found>
      InternalLayer = <local> <class 'returnn.tf.layers.base.InternalLayer'>
      sources = <not found>
  File "/mnt/projects/i6/returnn/returnn/tf/network.py", line 1077, in TFNetwork.add_layer
    line: layer = self._create_layer(name=name, layer_class=layer_class, **layer_desc)
    locals:
      layer = <not found>
      self = <local> <TFNetwork 'root/output(rec-subnet-output)' parent_layer=<RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      self._create_layer = <local> <bound method TFNetwork._create_layer of <TFNetwork 'root/output(rec-subnet-output)' parent_layer=<RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      name = <local> 'segments', len = 8
      layer_class = <local> <class 'returnn.tf.layers.base.InternalLayer'>
      layer_desc = <local> {'output': Data{'segments_gather_output', [T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]}, 'sources': []}
  File "/mnt/projects/i6/returnn/returnn/tf/network.py", line 1000, in TFNetwork._create_layer
    line: layer.output.sanity_check()
    locals:
      layer = <local> <InternalLayer output/'segments' out_type=Data{[T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]}>
      layer.output = <local> Data{'segments_gather_output', [T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]}
      layer.output.sanity_check = <local> <bound method Data.sanity_check of Data{'segments_gather_output', [T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]}>
  File "/mnt/projects/i6/returnn/returnn/tf/util/data.py", line 1487, in Data.sanity_check
    line: assert tag.dyn_size is not None
    locals:
      tag = <local> DimensionTag{'sliced-time:segments'[?]}
      tag.dyn_size = <local> None
AssertionError

Process finished with exit code 1
```

The error seems to happen after the construction of some `InternalLayer`. I tried to debug the construction and it seems that RETURNN is able to go through the `__init__` of `SliceNdLayer` and the `sanity_check` on the output is also working. However, later, another `InternalLayer` of `"segments"` is constructed, and there the `sanity_check` fails because the `dyn_size` of the output is not set. The error does not occur if I set `size` to an integer.

Reply:
Ah I see the problem. It uses the template from `layer_data_templates` (where the dyn size is unknown because this is just a template). I just made #652 which should fix that I think.

Reply:
Ok this is merged. Can you try again?

Reply:
Hm, I just did a rebase but the problems still persists:

```
/home/robin/miniconda3/envs/i6_env/bin/python3 /mnt/projects/i6/returnn/rnn.py /mnt/projects/i6/experiments/transducer/config/configs/segments-tests/error_reconstruct2.py
2021-09-14 15:34:34.439264: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2021-09-14 15:34:34.439287: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
RETURNN starting up, version 1.0.0+unknown, date/time 2021-09-14-15-34-35 (UTC+0200), pid 31609, cwd /mnt/projects/i6/returnn, Python /home/robin/miniconda3/envs/i6_env/bin/python3
RETURNN command line options: ['/mnt/projects/i6/experiments/transducer/config/configs/segments-tests/error_reconstruct2.py']
Hostname: robin-MS-7B86
TensorFlow: 2.5.0 (v2.5.0-rc3-213-ga4dfb8d1a71) (<site-package> in /home/robin/miniconda3/envs/i6_env/lib/python3.9/site-packages/tensorflow)
Setup TF inter and intra global thread pools, num_threads None, session opts {'log_device_placement': False, 'device_count': {'GPU': 0}}.
2021-09-14 15:34:35.823314: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-09-14 15:34:35.824436: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcuda.so.1
CUDA_VISIBLE_DEVICES is not set.
Collecting TensorFlow device list...
2021-09-14 15:34:35.864366: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-09-14 15:34:35.864386: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]      
2021-09-14 15:34:35.865298: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-09-14 15:34:35.865686: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1733] Found device 0 with properties: 
pciBusID: 0000:1f:00.0 name: NVIDIA GeForce GTX 1660 SUPER computeCapability: 7.5
coreClock: 1.8GHz coreCount: 22 deviceMemorySize: 5.80GiB deviceMemoryBandwidth: 312.97GiB/s
2021-09-14 15:34:35.865753: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2021-09-14 15:34:35.865791: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcublas.so.11'; dlerror: libcublas.so.11: cannot open shared object file: No such file or directory
2021-09-14 15:34:35.865826: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcublasLt.so.11'; dlerror: libcublasLt.so.11: cannot open shared object file: No such file or directory
2021-09-14 15:34:35.865862: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcufft.so.10'; dlerror: libcufft.so.10: cannot open shared object file: No such file or directory
2021-09-14 15:34:35.865895: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcurand.so.10'; dlerror: libcurand.so.10: cannot open shared object file: No such file or directory
2021-09-14 15:34:35.865929: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcusolver.so.11'; dlerror: libcusolver.so.11: cannot open shared object file: No such file or directory
2021-09-14 15:34:35.865962: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcusparse.so.11'; dlerror: libcusparse.so.11: cannot open shared object file: No such file or directory
2021-09-14 15:34:35.865996: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudnn.so.8'; dlerror: libcudnn.so.8: cannot open shared object file: No such file or directory
2021-09-14 15:34:35.866003: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1766] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2021-09-14 15:34:35.959372: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-09-14 15:34:35.959397: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]      0 
2021-09-14 15:34:35.959404: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1277] 0:   N 
Local devices available to TensorFlow:
  1/1: name: "/device:CPU:0"
       device_type: "CPU"
       memory_limit: 268435456
       locality {
       }
       incarnation: 16833894921831009270
Train data:
  input: 10 x 1
  output: {'classes': (3, 1), 'data': (10, 2)}
  DummyDataset, sequences: 10, frames: unknown
Device not set explicitly, and no GPU found.
Learning-rate-control: no file specified, not saving history (no proper restart possible)
Setup TF session with options {'log_device_placement': False, 'device_count': {'GPU': 0}} ...
2021-09-14 15:34:36.021195: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-09-14 15:34:36.021232: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]      
WARNING:tensorflow:From /mnt/projects/i6/returnn/returnn/tf/network.py:465: calling Zeros.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Warning: assuming dim tags are same with different size placeholders: DimensionTag{'time'[B?]} vs DimensionTag{'time'[B?]}
layer root/'data:data' output: Data{'data', [B,T|'time'[B],F|F'feature:data'(10)]}
layer root/'output' output: Data{'output_output', [T|'time'[B],B], dtype='int32', sparse=True, dim=3}
WARNING:tensorflow:From /mnt/projects/i6/returnn/returnn/tf/util/basic.py:1343: calling VarianceScaling.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
layer root/output(rec-subnet)/'data:classes' output: Data{'classes', [B], dtype='int32', sparse=True, dim=3, available_for_inference=False, ctx=loop('time'[B])}
layer root/output(rec-subnet)/'output' output: Data{'output_output', [B], dtype='int32', sparse=True, dim=3, ctx=loop('time'[B])}
layer root/output(rec-subnet)/'data:source' output: Data{'data', [B,F|F'feature:data'(10)], ctx=loop('time'[B])}
layer root/output(rec-subnet)/'output_prob' output: Data{'output_prob_output', [B,F|F'output_prob:feature-dense'(3)], ctx=loop('time'[B])}
layer root/output(rec-subnet)/'segment_starts' output: Data{'segment_starts_output', [B], dtype='int32', ctx=loop('time'[B])}
layer root/output(rec-subnet)/'segments' output: Data{'segments_gather_output', [B,T|'sliced-time:segments'[?],F|F'feature:data'(10)], ctx=loop('time'[B])}
WARNING:tensorflow:From /home/robin/miniconda3/envs/i6_env/lib/python3.9/site-packages/tensorflow/python/ops/array_ops.py:5043: calling gather (from tensorflow.python.ops.array_ops) with validate_indices is deprecated and will be removed in a future version.
Instructions for updating:
The `validate_indices` argument has no effect. Indices are always validated on CPU and never validated on GPU.
layer root/output(rec-subnet)/'output_is_not_blank' output: Data{'output_is_not_blank_output', [B], dtype='bool', sparse=True, dim=2, ctx=loop('time'[B])}
WARNING:tensorflow:From /mnt/projects/i6/returnn/returnn/tf/util/basic.py:5445: _colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
layer root/output(rec-subnet-output)/'segments' output: Data{'segments_gather_output', [T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]}
Exception creating layer root/output(rec-subnet-output)/'segments' of class InternalLayer with opts:
{'name': 'segments',
 'network': <TFNetwork 'root/output(rec-subnet-output)' parent_layer=<RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>,
 'output': Data{'segments_gather_output', [T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]},
 'sources': []}
Exception creating layer root/'output' of class RecLayer with opts:
{'_name': 'output',
 '_network': <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>,
 '_time_dim_tag': DimensionTag{'time'[B]},
 'back_prop': {'class': 'DummyDataset',
               'input_dim': 10,
               'num_seqs': 10,
               'output_dim': 3,
               'seq_len': 10},
 'include_eos': True,
 'n_out': <class 'returnn.util.basic.NotSpecified'>,
 'name': 'output',
 'network': <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>,
 'output': Data{'output_output', [T|'time'[B],B], dtype='int32', sparse=True, dim=3},
 'sources': [<SourceLayer 'data:data' out_type=Data{[B,T|'time'[B],F|F'feature:data'(10)]}>],
 'unit': <_SubnetworkRecCell 'root/output(rec-subnet)'>}
Unhandled exception <class 'AssertionError'> in thread <_MainThread(MainThread, started 140498237497728)>, proc 31609.

Thread current, main, <_MainThread(MainThread, started 140498237497728)>:
(Excluded thread.)

That were all threads.
EXCEPTION
Traceback (most recent call last):
  File "/mnt/projects/i6/returnn/rnn.py", line 11, in <module>
    line: main()
    locals:
      main = <local> <function main at 0x7fc7e5e34940>
  File "/mnt/projects/i6/returnn/returnn/__main__.py", line 655, in main
    line: execute_main_task()
    locals:
      execute_main_task = <global> <function execute_main_task at 0x7fc7e5e34820>
  File "/mnt/projects/i6/returnn/returnn/__main__.py", line 453, in execute_main_task
    line: engine.init_train_from_config(config, train_data, dev_data, eval_data)
    locals:
      engine = <global> <returnn.tf.engine.Engine object at 0x7fc7e5e52f70>
      engine.init_train_from_config = <global> <bound method Engine.init_train_from_config of <returnn.tf.engine.Engine object at 0x7fc7e5e52f70>>
      config = <global> <returnn.config.Config object at 0x7fc7ea52e820>
      train_data = <global> <DummyDataset 'train' epoch=None>
      dev_data = <global> None
      eval_data = <global> None
  File "/mnt/projects/i6/returnn/returnn/tf/engine.py", line 1031, in Engine.init_train_from_config
    line: self.init_network_from_config(config)
    locals:
      self = <local> <returnn.tf.engine.Engine object at 0x7fc7e5e52f70>
      self.init_network_from_config = <local> <bound method Engine.init_network_from_config of <returnn.tf.engine.Engine object at 0x7fc7e5e52f70>>
      config = <local> <returnn.config.Config object at 0x7fc7ea52e820>
  File "/mnt/projects/i6/returnn/returnn/tf/engine.py", line 1096, in Engine.init_network_from_config
    line: self._init_network(net_desc=net_dict, epoch=self.epoch)
    locals:
      self = <local> <returnn.tf.engine.Engine object at 0x7fc7e5e52f70>
      self._init_network = <local> <bound method Engine._init_network of <returnn.tf.engine.Engine object at 0x7fc7e5e52f70>>
      net_desc = <not found>
      net_dict = <local> {'output': {'class': 'rec', 'from': 'data:data', 'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, 'unit': {'segment_starts': {'class': 'switch', 'condition': 'prev:output_is_not_blank', 'true_from': ':i', 'false_from': '...
      epoch = <local> None
      self.epoch = <local> 1
  File "/mnt/projects/i6/returnn/returnn/tf/engine.py", line 1275, in Engine._init_network
    line: self.network, self.updater = self.create_network(
            config=self.config,
            extern_data=extern_data,
            rnd_seed=net_random_seed,
            train_flag=train_flag, eval_flag=self.use_eval_flag, search_flag=self.use_search_flag,
            initial_learning_rate=getattr(self, "initial_learning_rate", None),
            net_dict=net_desc)
    locals:
      self = <local> <returnn.tf.engine.Engine object at 0x7fc7e5e52f70>
      self.network = <local> None
      self.updater = <local> None
      self.create_network = <local> <bound method Engine.create_network of <class 'returnn.tf.engine.Engine'>>
      config = <not found>
      self.config = <local> <returnn.config.Config object at 0x7fc7ea52e820>
      extern_data = <local> <ExternData data={'data': Data{'data', [B,T|'time'[B],F|F'feature:data'(10)]}, 'classes': Data{'classes', [B,T|'time'[B]], dtype='int32', sparse=True, dim=3, available_for_inference=False}}>
      rnd_seed = <not found>
      net_random_seed = <local> 1
      train_flag = <local> <tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>
      eval_flag = <not found>
      self.use_eval_flag = <local> True
      search_flag = <not found>
      self.use_search_flag = <local> False
      initial_learning_rate = <not found>
      getattr = <builtin> <built-in function getattr>
      net_dict = <not found>
      net_desc = <local> {'output': {'class': 'rec', 'from': 'data:data', 'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, 'unit': {'segment_starts': {'class': 'switch', 'condition': 'prev:output_is_not_blank', 'true_from': ':i', 'false_from': '...
  File "/mnt/projects/i6/returnn/returnn/tf/engine.py", line 1316, in Engine.create_network
    line: network.construct_from_dict(net_dict)
    locals:
      network = <local> <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      network.construct_from_dict = <local> <bound method TFNetwork.construct_from_dict of <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      net_dict = <local> {'output': {'class': 'rec', 'from': 'data:data', 'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, 'unit': {'segment_starts': {'class': 'switch', 'condition': 'prev:output_is_not_blank', 'true_from': ':i', 'false_from': '...
  File "/mnt/projects/i6/returnn/returnn/tf/network.py", line 605, in TFNetwork.construct_from_dict
    line: self.construct_layer(net_dict, name, get_layer=get_layer)
    locals:
      self = <local> <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      self.construct_layer = <local> <bound method TFNetwork.construct_layer of <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      net_dict = <local> {'output': {'class': 'rec', 'from': 'data:data', 'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, 'unit': {'segment_starts': {'class': 'switch', 'condition': 'prev:output_is_not_blank', 'true_from': ':i', 'false_from': '...
      name = <local> 'output', len = 6
      get_layer = <local> None
  File "/mnt/projects/i6/returnn/returnn/tf/network.py", line 931, in TFNetwork.construct_layer
    line: return add_layer(name=name_with_prefix, layer_class=layer_class, **layer_desc)
    locals:
      add_layer = <local> <bound method TFNetwork.add_layer of <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      name = <local> 'output', len = 6
      name_with_prefix = <local> 'output', len = 6
      layer_class = <local> <class 'returnn.tf.layers.rec.RecLayer'>
      layer_desc = <local> {'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, '_network': <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>, '_name': 'output', 'n_out': <class 'returnn.util.basic.NotSpecified'>, 'sourc..., len = 8
  File "/mnt/projects/i6/returnn/returnn/tf/network.py", line 1077, in TFNetwork.add_layer
    line: layer = self._create_layer(name=name, layer_class=layer_class, **layer_desc)
    locals:
      layer = <not found>
      self = <local> <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      self._create_layer = <local> <bound method TFNetwork._create_layer of <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      name = <local> 'output', len = 6
      layer_class = <local> <class 'returnn.tf.layers.rec.RecLayer'>
      layer_desc = <local> {'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, '_network': <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>, '_name': 'output', 'n_out': <class 'returnn.util.basic.NotSpecified'>, 'sourc..., len = 8
  File "/mnt/projects/i6/returnn/returnn/tf/network.py", line 998, in TFNetwork._create_layer
    line: layer = layer_class(**layer_desc)
    locals:
      layer = <not found>
      layer_class = <local> <class 'returnn.tf.layers.rec.RecLayer'>
      layer_desc = <local> {'include_eos': True, 'back_prop': {'class': 'DummyDataset', 'input_dim': 10, 'output_dim': 3, 'num_seqs': 10, 'seq_len': 10}, '_network': <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>, '_name': 'output', 'n_out': <class 'returnn.util.basic.NotSpecified'>, 'sourc..., len = 11
  File "/mnt/projects/i6/returnn/returnn/tf/layers/rec.py", line 238, in RecLayer.__init__
    line: y = self._get_output_subnet_unit(self.cell)
    locals:
      y = <not found>
      self = <local> <RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}>
      self._get_output_subnet_unit = <local> <bound method RecLayer._get_output_subnet_unit of <RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}>>
      self.cell = <local> <_SubnetworkRecCell 'root/output(rec-subnet)'>
  File "/mnt/projects/i6/returnn/returnn/tf/layers/rec.py", line 925, in RecLayer._get_output_subnet_unit
    line: output = cell.get_output()
    locals:
      output = <not found>
      cell = <local> <_SubnetworkRecCell 'root/output(rec-subnet)'>
      cell.get_output = <local> <bound method _SubnetworkRecCell.get_output of <_SubnetworkRecCell 'root/output(rec-subnet)'>>
  File "/mnt/projects/i6/returnn/returnn/tf/layers/rec.py", line 2560, in _SubnetworkRecCell.get_output
    line: self._construct_output_layers_moved_out(
            loop_accumulated=self.final_acc_tas_dict, seq_len=seq_len,
            extra_output_layers=extra_output_layers, final_net_vars=final_net_vars)
    locals:
      self = <local> <_SubnetworkRecCell 'root/output(rec-subnet)'>
      self._construct_output_layers_moved_out = <local> <bound method _SubnetworkRecCell._construct_output_layers_moved_out of <_SubnetworkRecCell 'root/output(rec-subnet)'>>
      loop_accumulated = <not found>
      self.final_acc_tas_dict = <local> {'loss_output_prob': <tf.TensorArray 'output/rec/subnet_base/acc_ta_loss_output_prob'>, 'error_output_prob': <tf.TensorArray 'output/rec/subnet_base/acc_ta_error_output_prob'>, 'output_output': <tf.TensorArray 'output/rec/subnet_base/acc_ta_output_output'>, 'output_segments': <tf.TensorArray 'out...
      seq_len = <local> <tf.Tensor 'output/rec/subnet_base/check_seq_len_batch_size/check_input_dim/identity_with_dim_check:0' shape=(?,) dtype=int32>
      extra_output_layers = <local> {'segments', 'output'}, len = 2
      final_net_vars = <local> ([<tf.Tensor 'output/rec/while/Exit_1:0' shape=(?,) dtype=bool>, <tf.Tensor 'output/rec/while/Exit_2:0' shape=(?,) dtype=int32>], [])
  File "/mnt/projects/i6/returnn/returnn/tf/layers/rec.py", line 3322, in _SubnetworkRecCell._construct_output_layers_moved_out
    line: self.output_layers_net.layers[layer_name] = get_layer(layer_name)
    locals:
      self = <local> <_SubnetworkRecCell 'root/output(rec-subnet)'>
      self.output_layers_net = <local> <TFNetwork 'root/output(rec-subnet-output)' parent_layer=<RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      self.output_layers_net.layers = <local> {}
      layer_name = <local> 'segments', len = 8
      get_layer = <local> <function _SubnetworkRecCell._construct_output_layers_moved_out.<locals>.get_layer at 0x7fc84841cc10>
  File "/mnt/projects/i6/returnn/returnn/tf/layers/rec.py", line 3314, in _SubnetworkRecCell._construct_output_layers_moved_out.<locals>.get_layer
    line: return get_loop_acc_layer(name)
    locals:
      get_loop_acc_layer = <local> <function _SubnetworkRecCell._construct_output_layers_moved_out.<locals>.get_loop_acc_layer at 0x7fc8483cb820>
      name = <local> 'segments', len = 8
  File "/mnt/projects/i6/returnn/returnn/tf/layers/rec.py", line 3256, in _SubnetworkRecCell._construct_output_layers_moved_out.<locals>.get_loop_acc_layer
    line: layer_ = self.output_layers_net.add_layer(
            name=name, output=output, layer_class=InternalLayer, sources=[])
    locals:
      layer_ = <not found>
      self = <local> <_SubnetworkRecCell 'root/output(rec-subnet)'>
      self.output_layers_net = <local> <TFNetwork 'root/output(rec-subnet-output)' parent_layer=<RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      self.output_layers_net.add_layer = <local> <bound method TFNetwork.add_layer of <TFNetwork 'root/output(rec-subnet-output)' parent_layer=<RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      name = <local> 'segments', len = 8
      output = <local> Data{'segments_gather_output', [T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]}
      layer_class = <not found>
      InternalLayer = <local> <class 'returnn.tf.layers.base.InternalLayer'>
      sources = <not found>
  File "/mnt/projects/i6/returnn/returnn/tf/network.py", line 1077, in TFNetwork.add_layer
    line: layer = self._create_layer(name=name, layer_class=layer_class, **layer_desc)
    locals:
      layer = <not found>
      self = <local> <TFNetwork 'root/output(rec-subnet-output)' parent_layer=<RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      self._create_layer = <local> <bound method TFNetwork._create_layer of <TFNetwork 'root/output(rec-subnet-output)' parent_layer=<RecLayer 'output' out_type=Data{[T|'time'[B],B], dtype='int32', sparse=True, dim=3}> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      name = <local> 'segments', len = 8
      layer_class = <local> <class 'returnn.tf.layers.base.InternalLayer'>
      layer_desc = <local> {'output': Data{'segments_gather_output', [T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]}, 'sources': []}
  File "/mnt/projects/i6/returnn/returnn/tf/network.py", line 1000, in TFNetwork._create_layer
    line: layer.output.sanity_check()
    locals:
      layer = <local> <InternalLayer output/'segments' out_type=Data{[T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]}>
      layer.output = <local> Data{'segments_gather_output', [T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]}
      layer.output.sanity_check = <local> <bound method Data.sanity_check of Data{'segments_gather_output', [T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]}>
  File "/mnt/projects/i6/returnn/returnn/tf/util/data.py", line 1487, in Data.sanity_check
    line: assert tag.dyn_size is not None
    locals:
      tag = <local> DimensionTag{'sliced-time:segments'[?]}
      tag.dyn_size = <local> None
AssertionError

Process finished with exit code 1
```

Reply:
Which layer do you mean? The `layer = <local> <InternalLayer output/'segments' out_type=Data{[T|'time'[B],B,'sliced-time:segments'[?],F|F'feature:data'(10)]}>`?

Reply:
No, the change I did via #652, in `get_loop_acc_layer`.

Reply:
Oh okay. Can I look at this tomorrow?

Reply:
Yea sure, this is your PR anyway.

Reply:
So I just did a lot of debugging and I think I found the root of the problem. In our `__init__` we set `slice_tag` to the tag which we created in `get_out_data_from_opts`. This tag has its `same_as` attribute set to itself. Therefore, even after we set the `dyn_size` of `slice_tag`, the `same_as` attribute is still the tag which does not have `dyn_size` set. Then later, in `get_loop_acc_layer`, the `in_loop_layer.output` has the same tag as `slice_tag` for its slice axis. However, during the call of `copy_template_set_ctx` in `get_loop_acc_layer`, a new `Data` object is constructed and inside the `Data.__init__` inside `self._adapt_batch_consistent_dim_tags()`, the tag of the slice axis is set to its `same_as` attribute, which does not have the `dyn_size` set and therefore the exception is raised during the sanity check.

To fix it, I looked at the previous `SliceNdLayer` implementation, which doesn't throw the error. Here, in the `__init__`, a new dimension tag is created instead of copying the one from `get_out_data_from_opts` and the `output.size_placeholder` is also set explicitly.
1. I first tried to just set the `same_as` attribute of `slice_tag` to `None`. This led to the same error.
2. I then tried to create an entirely new dimension tag for `slice_tag`, like it is done in the original implemenation of the layer, but this also led to the same error.
3. I then additionally tried to also manually set `self.output.size_placeholder[start_data.ndim] = dyn_size` and this worked.

The question is how we continue from here. My solution seems hacky because I guess we would like to use the `size_placeholder` from `GatherLayer` instead of setting it manually. What do you think

Reply:
> This tag has its `same_as` attribute set to itself.

I don't understand. So `self.same_as is self`? This sounds wrong.

Reply:
Note that `dyn_size` is a `property`, and its assignment has some extra logic. It sounds like this extra logic maybe behaves wrong.

Also `size_placeholder` assignments have extra logic. `size_placeholder` assignments anyway just wrap to `set_dynamic_size`, and that one does the extra logic.

`dyn_size_ext` assignments do not have extra logic.

This is not really consistent, and likely the reason for the problems, and why it works in one case, and not in the other.

Reply:
I will look a bit into this, how we can unify this to have the same logic consistently everywhere, or whether this is possible. `size_placeholder` assignments have this special logic on purpose to support older code.

Or maybe you or @Zettelkasten have also some suggestions on this?

Reply:
> > This tag has its `same_as` attribute set to itself.
> 
> I don't understand. So `self.same_as is self`? This sounds wrong.

No no, after we created the tag in `get_out_data_from_opts`, we do `gather_positions_data = gather_positions_data.copy_add_dim_by_tag(tag, unbroadcast=True, axis=start_data.batch_ndim)` and this sets `same_as` to the same dimension tag as itself. See https://github.com/rwth-i6/returnn/blob/c2d69b528550bd4af07d5afcf135bcab2e6e9bdf/returnn/tf/util/data.py#L429, which is called during `copy_add_dim_by_tag`. It's a bit confusing to put into words.

Reply:
> Or maybe you or @Zettelkasten have also some suggestions on this?

I don't have much time today but maybe I can find a non-hacky way to make it work

Reply:
I still don't understand.

> this sets `same_as` to the same dimension tag as itself

What is "itself" here? So `tag.same_as is tag`?

This tag `tag` from `get_out_data_from_opts` is the same instance as `slice_tag` in `__init__`?

Reply:
Ah I think I understand. The problem is only due to `copy_template_set_ctx` in `get_loop_acc_layer`? All other logic is fine, esp in `SliceNdLayer`?

Reply:
> Ah I think I understand. The problem is only due to `copy_template_set_ctx` in `get_loop_acc_layer`? All other logic is fine, esp in `SliceNdLayer`?

Yes, so before `copy_template_set_ctx`, it calls `copy_template_adding_time_dim` and this returns a `Data` object where the `dyn_size` of the slice axis is set. But then, `copy_template_set_ctx` returns a `Data` object, where the `dyn_size` of the slice axis is set to `None` and this leads to the error

Reply:
I think the error is actually correct (probably just misleading or not really intuitive).

When you have different dyn sizes of shape [B] inside the loop per each loop frame, then the dyn size outside the loop should be of shape [B,T]. To differentiate those cases (inside loop vs outside loop), we have this context on dim tags and `Data`.

In `get_loop_acc_layer`, it correctly assigns the needed context (outside the loop). When the dyn size was set only for within the loop context, it is unknown outside. Or we can infer it by accumulating it as well, but this is currently not implemented.

The question is also, is it actually correct that you have a different dyn size per each loop frame?

Reply:
Ok, from your other comment, you say it is correct that you have a different dyn size (shape [B]) per each loop frame.
That implies that the dyn size outside the loop would be of shape [B,T].

This is currently not implemented. The implementation would need to accumulate the dyn sizes.

The error then is correct in that the dyn size outside the loop is not known.

So what's missing is this implementation. I will start working on this now.

Reply:
> The question is also, is it actually correct that you have a different dyn size per each loop frame?

Yes, so I can give a simple example:

imagine we have batch size 2 and the following sequences:
`B1BB2`
`2B3BB`
where `B`, `1` and `2` are some class labels. Then, in my reccurrent unit I would like to get the following sequences:

In the first rec step:
`B`
`2`
In the second step:
`B1`
`B`
In the third step:
`B`
`B3`
In the fourth step:
`BB`
`B`
In the fifth step:
`BB2`
`BB`
and so on. So I want to gather the values from the current rec step until the previous number. This leads to different dynamic sizes per each loop frame.

Reply:
> So what's missing is this implementation. I will start working on this now.

Okay, thank you. I sadly don't have any more time today but I will look into this again tomorrow and fix the other errors which you commented on.

Reply:
I started the PR #657 for this. So far I'm still working only on the test case, which is also not so trivial. So still WIP. I'm also not sure how much more time I have today.

Reply:
Btw, this problem only occurs because you have `optimize_move_layers_out=False` and also `"is_output_layer": True` on the `segments` layer. Why do you have the latter actually?

I wonder whether we need that right now for the PR. I think we can already proceed with this PR here and skip this aspect, or not?

Reply:
> Btw, this problem only occurs because you have `optimize_move_layers_out=False` and also `"is_output_layer": True` on the `segments` layer. Why do you have the latter actually?

I set `optimize_move_layers_out=False` to see what happens during inference, when the loop cannot be optimized. And if I don't set `"is_output_layer": True`, then RETURNN does not construct the layer because it is not used otherwise and therefore the error does not occur.

Reply:
For your test, you could just put another reduce layer afterwards, which reduces over the new dim. The problem is really that this is specifically an output layer. In principle, this should also work, but this is not implemented yet, but also not needed for your use case.

Reply:
> For your test, you could just put another reduce layer afterwards, which reduces over the new dim

Perfect, this works!